File: LinkedList/le206_reverse_linkedlist.py
"""
Reverse a singly linked list.
"""
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

    
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        cur = head  # this is more clear 
        prev = None
        while cur!= None:
            temp = cur.next
            cur.next = prev
            prev = cur
            cur = temp
        return prev

File: LinkedList/le206_reverseLinkedList.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode reverseList(ListNode head) {
        ListNode prev = null;
        while (head != null) {
            ListNode temp = head.next;
            head.next = prev;
            prev = head;
            head = temp;
        }
        return prev;
    }
}


public class Solution2 {  //recursion
    public ListNode reverseList(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode newnode = reverseList(head.next);
        head.next.next = head;
        head.next = null;
        return newnode;
    }
}
